### PR TITLE
Feat: sebulba rec ippo

### PR DESCRIPTION
**Pull request by SimonDuToit**
Sebulba implementation of recurrent IPPO.

### CODE REVIEW

File: mava/configs/default/rec_ippo_sebulba.yaml [hunk 1]
```diff
@@ -0,0 +1,11 @@
+defaults:
+  - logger: logger
+  - arch: sebulba
+  - system: ppo/rec_ippo
+  - network: rnn  # [mlp, continuous_mlp, cnn]
+  - env: lbf_gym  # [rware_gym, lbf_gym]
```
Comment (OmaymaMahjoub) [inline mava/configs/default/rec_ippo_sebulba.yaml]: if you can add smaclite_gym to the list

File: mava/systems/ppo/sebulba/rec_ippo.py [hunk 2]
```diff
@@ -0,0 +1,910 @@
...
+    num_agents, num_envs = config.system.num_agents, config.arch.num_envs
+    move_to_device = lambda x: jax.device_put(x, device=actor_device)
+
+    @jax.jit
+    def act_fn(
+        params: Params,
+        observation: Observation,
+        dones,
+        hstates,
+        key: chex.PRNGKey,
```
Comment (OmaymaMahjoub) [inline mava/systems/ppo/sebulba/rec_ippo.py:107]:
```suggestion
        dones: chex.Array,
        hstates: HiddenStates,
```

File: mava/systems/ppo/sebulba/rec_ippo.py [hunk 3]
```diff
@@ -0,0 +1,910 @@
+# Copyright 2022 InstaDeep Ltd. All rights reserved.
```
Comment (OmaymaMahjoub) [inline mava/systems/ppo/sebulba/rec_ippo.py:1]: If you can update the typings in Pipeline `mava/utils/sebulba.py` to be `Union[PPOTransition, RNNPPOTransition]`

File: mava/systems/ppo/sebulba/rec_ippo.py [hunk 4]
```diff
@@ -0,0 +1,910 @@
...
+        )
+        critic_hidden_state, value = critic_apply_fn(
+            params.critic_params, hstates.critic_hidden_state, ac_in
+        )
+
+        action = actor_policy.sample(seed=key)
+        log_prob = actor_policy.log_prob(action)
+        # It may be faster to calculate the values in the learner as
+        # then we won't need to pass critic params to actors.
+        # value = critic_apply_fn(params.critic_params, observation).squeeze()
```
Comment (OmaymaMahjoub) [inline mava/systems/ppo/sebulba/rec_ippo.py]: if you can remove this comment

File: mava/systems/ppo/sebulba/rec_ippo.py [hunk 5]
```diff
@@ -0,0 +1,910 @@
...
+        hstates = HiddenStates(policy_hidden_state, critic_hidden_state)
+        return action, log_prob, value, hstates
+
+    timestep = env.reset(seed=seeds)
+    dones = np.repeat(timestep.last(), num_agents).reshape(num_envs, -1)
+
+    # simon
```
Comment (OmaymaMahjoub) [inline mava/systems/ppo/sebulba/rec_ippo.py]: same here once you are done from cleaning, if you can remove this comment

Comment (OmaymaMahjoub) [reply mava/systems/ppo/sebulba/rec_ippo.py]: I see this comment exists in multiple lines, if you can remove them all

File: mava/systems/ppo/sebulba/rec_ippo.py [hunk 6]
```diff
@@ -0,0 +1,910 @@
...
+            return update_state, loss_info
+
+        update_state = (params, opt_states, traj_batch, advantages, targets, key)
+        # Update epochs
+        update_state, loss_info = jax.lax.scan(
+            _update_epoch, update_state, None, config.system.ppo_epochs
+        )
+
+        params, opt_states, traj_batch, advantages, targets, key = update_state
+        # learner_state = LearnerState(params, opt_states, key, None, learner_state.timestep)
```
Comment (OmaymaMahjoub) [inline mava/systems/ppo/sebulba/rec_ippo.py]: If you can remove this comment

**Review by OmaymaMahjoub — CHANGES_REQUESTED** [review]
Overall the system looks correct and reasonable. Well done Simon! I just kept minor requests :)

File: mava/systems/ppo/sebulba/rec_ippo.py (hunk 3 above)
Comment (SimonDuToit) [reply mava/systems/ppo/sebulba/rec_ippo.py:1]: This causes errors in the pre-commit. For now I changed both sebulba systems to use the `MavaTransition` type-var but this is probably a temporary solution.

Comment (sash-a) [reply mava/systems/ppo/sebulba/rec_ippo.py:1]: Can you please make an issue for this. I think the best solution is to make a protocol with the all the common things in a transition (actions, obs, done, reward). The challenge is that named tuples don't seem to work with protocols so we'd likely need to switch to a flax/chex dataclass

File: mava/systems/ppo/sebulba/rec_ippo.py [hunk 7]
```diff
@@ -0,0 +1,898 @@
...
+        # Squeeze out the batch dimension and mask out the value of terminal states.
+        last_val = last_val.squeeze(0)
+        advantages, targets = _calculate_gae(traj_batch, last_val, last_done)
+
+        def _update_epoch(update_state: Tuple, _: Any) -> Tuple[Tuple, Metrics]:
+            """Update the network for a single epoch."""
+
+            def _update_minibatch(train_state: Tuple, batch_info: Tuple) -> Tuple:
+                """Update the network for a single minibatch."""
+                # UNPACK TRAIN STATE AND BATCH INFO
```
Comment (sash-a) [inline mava/systems/ppo/sebulba/rec_ippo.py]: In line with PPO updates I did last year

```suggestion
```

File: mava/systems/ppo/sebulba/rec_ippo.py [hunk 8]
```diff
@@ -0,0 +1,898 @@
...
+                traj_batch, advantages, targets = batch_info
+
+                def _actor_loss_fn(
+                    actor_params: FrozenDict,
+                    traj_batch: RNNPPOTransition,
+                    gae: chex.Array,
+                    key: chex.PRNGKey,
+                ) -> Tuple:
+                    """Calculate the actor loss."""
+                    # RERUN NETWORK
```
Comment (sash-a) [inline mava/systems/ppo/sebulba/rec_ippo.py]:
```suggestion
                    # Rerun network
```

File: mava/systems/ppo/sebulba/rec_ippo.py [hunk 9]
```diff
@@ -0,0 +1,898 @@
...
+                    total_loss = loss_actor - config.system.ent_coef * entropy
+                    return total_loss, (loss_actor, entropy)
+
+                def _critic_loss_fn(
+                    critic_params: FrozenDict,
+                    traj_batch: RNNPPOTransition,
+                    targets: chex.Array,
+                ) -> Tuple:
+                    """Calculate the critic loss."""
+                    # RERUN NETWORK
```
Comment (sash-a) [inline mava/systems/ppo/sebulba/rec_ippo.py]:
```suggestion
                    # Rerun network
```

File: mava/systems/ppo/sebulba/rec_ippo.py [hunk 10]
```diff
@@ -0,0 +1,898 @@
...
+                    obs_and_done = (traj_batch.obs, traj_batch.done)
+                    _, value = critic_apply_fn(
+                        critic_params, traj_batch.hstates.critic_hidden_state[0], obs_and_done
+                    )
+
+                    # CALCULATE VALUE LOSS
```
Comment (sash-a) [inline mava/systems/ppo/sebulba/rec_ippo.py]:
```suggestion
                    # Calculate value loss
```

File: mava/systems/ppo/sebulba/rec_ippo.py [hunk 11]
```diff
@@ -0,0 +1,898 @@
...
+                        )
+                        * gae
+                    )
+                    loss_actor = -jnp.minimum(loss_actor1, loss_actor2)
+                    loss_actor = loss_actor.mean()
+                    # The seed will be used in the TanhTransformedDistribution:
+                    entropy = actor_policy.entropy(seed=key).mean()
+
+                    total_loss = loss_actor - config.system.ent_coef * entropy
+                    return total_loss, (loss_actor, entropy)
```
Comment (sash-a) [inline mava/systems/ppo/sebulba/rec_ippo.py]:
```suggestion
                    actor_loss1 = ratio * gae
                    actor_loss2 = (
                        jnp.clip(
                            ratio,
                            1.0 - config.system.clip_eps,
                            1.0 + config.system.clip_eps,
                        )
                        * gae
                    )
                    actor_loss = -jnp.minimum(actor_loss1, actor_loss2)
                    actor_loss = actor_loss.mean()
                    # The seed will be used in the TanhTransformedDistribution:
                    entropy = actor_policy.entropy(seed=key).mean()

                    total_loss = actor_loss - config.system.ent_coef * entropy
                    return total_loss, (actor_loss, entropy)
```

File: mava/systems/ppo/sebulba/rec_ippo.py [hunk 12]
```diff
@@ -0,0 +1,898 @@
...
+    learn = jax.jit(
+        shard_map(
+            learn,
+            mesh=mesh,
+            in_specs=(learn_state_spec, data_spec),
+            out_specs=(learn_state_spec, data_spec),
+        )
+    )
+
+    # # Load model from checkpoint if specified.
```
Comment (sash-a) [inline mava/systems/ppo/sebulba/rec_ippo.py]:
```suggestion
    # Load model from checkpoint if specified.
```

File: mava/systems/ppo/sebulba/rec_ippo.py [hunk 13]
```diff
@@ -0,0 +1,898 @@
...
+                    # CALCULATE VALUE LOSS
+                    value_pred_clipped = traj_batch.value + (value - traj_batch.value).clip(
+                        -config.system.clip_eps, config.system.clip_eps
+                    )
+                    value_losses = jnp.square(value - targets)
+                    value_losses_clipped = jnp.square(value_pred_clipped - targets)
+                    value_loss = 0.5 * jnp.maximum(value_losses, value_losses_clipped).mean()
+
+                    total_loss = config.system.vf_coef * value_loss
+                    return total_loss, value_loss
```
Comment (sash-a) [inline mava/systems/ppo/sebulba/rec_ippo.py:339]:
```suggestion
                    total_value_loss = config.system.vf_coef * value_loss
                    return total_value_loss, value_loss
```

File: mava/systems/ppo/sebulba/rec_ippo.py [hunk 14]
```diff
@@ -0,0 +1,898 @@
...
+                # Calculate actor loss
+                key, entropy_key = jax.random.split(key)
+                actor_grad_fn = jax.value_and_grad(_actor_loss_fn, has_aux=True)
+                actor_loss_info, actor_grads = actor_grad_fn(
+                    params.actor_params, traj_batch, advantages, entropy_key
+                )
+
+                # Calculate critic loss
+                critic_grad_fn = jax.value_and_grad(_critic_loss_fn, has_aux=True)
+                critic_loss_info, critic_grads = critic_grad_fn(
```
Comment (sash-a) [inline mava/systems/ppo/sebulba/rec_ippo.py]:
```suggestion
                value_loss_info, critic_grads = critic_grad_fn(
```

File: mava/systems/ppo/sebulba/rec_ippo.py [hunk 15]
```diff
@@ -0,0 +1,898 @@
...
+                # pmean over learner devices.
+                actor_grads, actor_loss_info = jax.lax.pmean(
+                    (actor_grads, actor_loss_info),
+                    axis_name="learner_devices",
+                )
+
+                # pmean over learner devices.
+                critic_grads, critic_loss_info = jax.lax.pmean(
+                    (critic_grads, critic_loss_info), axis_name="learner_devices"
+                )
```
Comment (sash-a) [inline mava/systems/ppo/sebulba/rec_ippo.py]:
```suggestion
                critic_grads, value_loss_info = jax.lax.pmean(
                    (critic_grads, value_loss_info), axis_name="learner_devices"
                )
```

File: mava/systems/ppo/sebulba/rec_ippo.py [hunk 16]
```diff
@@ -0,0 +1,898 @@
...
+                # Pack loss info
+                actor_total_loss, (actor_loss, entropy) = actor_loss_info
+                critic_total_loss, (value_loss) = critic_loss_info
+                total_loss = critic_total_loss + actor_total_loss
+                loss_info = {
+                    "total_loss": total_loss,
+                    "value_loss": value_loss,
+                    "actor_loss": actor_loss,
+                    "entropy": entropy,
+                }
```
Comment (sash-a) [inline mava/systems/ppo/sebulba/rec_ippo.py]:
```suggestion
                actor_loss, (_, entropy) = actor_loss_info
                value_loss, (unscaled_value_loss) = value_loss_info
                total_loss = actor_loss + value_loss
                loss_info = {
                    "total_loss": total_loss,
                    "value_loss": unscaled_value_loss,
                    "actor_loss": actor_loss,
                    "entropy": entropy,
                }
```

File: mava/systems/ppo/sebulba/rec_ippo.py [hunk 17]
```diff
@@ -0,0 +1,898 @@
...
+            # permutation = jax.random.permutation(shuffle_key, batch_size)
+            batch = (traj_batch, advantages, targets)
+
+            num_recurrent_chunks = (
+                config.system.rollout_length // config.system.recurrent_chunk_size
+            )
+            batch = tree.map(
+                lambda x: x.reshape(
+                    config.system.recurrent_chunk_size,
+                    num_learner_envs * num_recurrent_chunks,
```
Comment (sash-a) [inline mava/systems/ppo/sebulba/rec_ippo.py]: Can you extract this into a variable called `batch_size`. Will make it more clear and it's also used below

File: mava/systems/ppo/sebulba/rec_ippo.py [hunk 18]
```diff
@@ -0,0 +1,898 @@
...
+
+            params, opt_states, traj_batch, advantages, targets, key = update_state
+
+            key = jnp.squeeze(key, axis=0)  # Remove the learner_devices axis
+            key, shuffle_key, entropy_key = jax.random.split(key, 3)
+            key = jnp.expand_dims(key, axis=0)  # add the learner_devices axis for shape consitency
+
+            # Shuffle minibatches
+            # batch_size = config.system.rollout_length * num_learner_envs
+            # permutation = jax.random.permutation(shuffle_key, batch_size)
```
Comment (sash-a) [inline mava/systems/ppo/sebulba/rec_ippo.py]:
```suggestion
```

**Review by sash-a — CHANGES_REQUESTED** [review]
Looks great! Pretty much good to go except a few minor style changes to bring it up to date with the latest PPO changes that went in at the end of last year

File: mava/systems/ppo/sebulba/rec_ippo.py [hunk 19]
```diff
@@ -0,0 +1,897 @@
...
+
+    # Initialise hidden states.
+    init_policy_hstate = ScannedRNN.initialize_carry(
+        (config.arch.num_envs, num_agents), config.network.hidden_state_dim
+    )
+    init_critic_hstate = ScannedRNN.initialize_carry(
+        (config.arch.num_envs, num_agents), config.network.hidden_state_dim
+    )
+    hstates = HiddenStates(init_policy_hstate, init_critic_hstate)
+    hstates_tpu = tree.map(move_to_device, hstates)
```
Comment (sash-a) [inline mava/systems/ppo/sebulba/rec_ippo.py]: `device_put` already tree maps
```suggestion
    hstates_tpu = move_to_device(hstates)
```

File: mava/systems/ppo/sebulba/rec_ippo.py [hunk 20]
```diff
@@ -0,0 +1,897 @@
...
+        traj: List[RNNPPOTransition] = []
+        episode_metrics: List[Dict] = []
+        actor_timings: Dict[str, List[float]] = defaultdict(list)
+        with RecordTimeTo(actor_timings["rollout_time"]):
+            for _ in range(config.system.rollout_length):
+                with RecordTimeTo(actor_timings["get_params_time"]):
+                    params = params_source.get()  # Get the latest parameters from the learner
+
+                obs_tpu = tree.map(move_to_device, timestep.observation)
+                last_dones = tree.map(move_to_device, dones)
```
Comment (sash-a) [inline mava/systems/ppo/sebulba/rec_ippo.py]:
```suggestion
                obs_tpu = move_to_device(timestep.observation)
                last_dones = move_to_device(dones)
```

**Review by sash-a — COMMENTED** [review]
Final minor comments
